tests/test_voc_format.py

```diff
@@ -496,6 +537,228 @@ def test_can_pickle(self):
                 compare_datasets_strict(self, source, parsed)
 
 
+class VocExtractorTest(TestCase):
+    # ?xml... must be in the file beginning
+    XML_ANNOTATION_TEMPLATE = """<?xml version="1.0" encoding="UTF-8"?>
+<annotation>
+<filename>a.jpg</filename>
+<size><width>20</width><height>10</height><depth>3</depth></size>
+<object>
+    <name>cat</name>
+    <bndbox><xmin>1</xmin><ymin>2</ymin><xmax>3</xmax><ymax>4</ymax></bndbox>
+    <difficult>1</difficult>
+    <truncated>1</truncated>
+    <occluded>1</occluded>
+</object>
+</annotation>
+    """
+
+    @classmethod
+    def _write_xml_dataset(cls, root_dir, fmt_dir="Main", mangle_list_file=None, mangle_xml=None):
+        subset_file = osp.join(root_dir, "ImageSets", fmt_dir, "test.txt")
+        os.makedirs(osp.dirname(subset_file))
+        with open(subset_file, "w") as f:
+            text = "a\n" if fmt_dir != "Layout" else "a 0\n"
+            if mangle_list_file:
+                text = mangle_list_file(text)
+            f.write(text)
+
+        ann_file = osp.join(root_dir, "Annotations", "a.xml")
+        os.makedirs(osp.dirname(ann_file))
+        with open(ann_file, "w") as f:
+            text = cls.XML_ANNOTATION_TEMPLATE
+            if mangle_xml:
+                text = mangle_xml(text)
```

Comment:
IMO, mangling the XML content as _text_ leads to less clear and less robust tests. For example, there's a test below that matches for `<name>cat</name>`, but there's no guarantee that the match will happen inside the object. So if the template is later extended to contain multiple instances of `<name>cat</name>`, then the test might end up doing something unexpected.

It would be better to parse the XML file here, give the `mangle_xml` function the root element, then dump it back to XML. Then the mangle function could look like this:

```python
def mangle_xml(root):
    root.find('object/name').text = 'test'
```